spatialprofilingtoolbox/ondemand/providers/counts_provider.py

```diff
@@ -0,0 +1,103 @@
+class CountsProvider(Provider):
+    """Scan binary-format expression matrices for specific signatures."""
+
+    def __init__(self, data_directory: str, load_centroids: bool = False) -> None:
+        """Load from a precomputed JSON artifact in the data directory.
+
+        Note: CountsProvider never loads centroids because it does not need them.
+        """
+        super().__init__(data_directory, False)
+
+    def compute_signature(self, channel_names: list[str], study_name: str) -> int:
+        """Compute int signature of this channel name combination."""
+        target_by_symbol = self.studies[study_name]['target by symbol']
+        target_index_lookup = self.studies[study_name]['target index lookup']
+        if not all(name in target_by_symbol.keys() for name in channel_names):
+            return None
+        identifiers = [target_by_symbol[name] for name in channel_names]
+        indices = [target_index_lookup[identifier]
+                   for identifier in identifiers]
+        signature = 0
+        for index in indices:
+            signature = signature + (1 << index)
+        return signature
+
+    def count_structures_of_exact_signature(
+        self,
+        signature: int,
+        study_name: str,
+    ) -> dict[str, list[int]]:
+        """Count the number of structures that have this exact channel signature."""
+        counts = {}
+        for specimen, data_array in self.data_arrays[study_name].items():
+            count = (data_array['entry'] == signature).sum()
+            # count = 0
+            # for entry in data_array:
+            #     if entry == signature:
+            #         count = count + 1
+            counts[specimen] = [count, len(data_array)]
+        return counts
+
+    def count_structures_of_partial_signature(
+        self,
+        signature: int,
+        study_name: str,
+    ) -> dict[str, list[int]]:
+        """Count the number of structures that have part of this channel signature."""
+        counts = {}
+        for specimen, data_array in self.data_arrays[study_name].items():
+            count = 0
+            for entry in data_array['entry']:
+                if entry | signature == entry:
+                    count = count + 1
+            counts[specimen] = [count, len(data_array)]
+        return counts
+
+    def count_structures_of_partial_signed_signature(
+        self,
+        positives_signature: int,
+        negatives_signature: int,
+        study_name: str,
+    ) -> dict[str, list[int]]:
+        """Count the number of structures that part of this signature, in signs."""
+        counts = {}
+        for specimen, data_array in self.data_arrays[study_name].items():
+            count = 0
+            for entry in data_array['entry']:
+                if (entry | positives_signature == entry) and \
+                        (~entry | negatives_signature == ~entry):
+                    count = count + 1
+            counts[specimen] = [count, len(data_array)]
```

Comment:
This is always [0, 0] in tests.

Reply:
I think this line is the culprit:

https://github.com/nadeemlab/SPT/blob/d38166e5a1a5e11d8ffdbc486e74a8db87bb6d53/spatialprofilingtoolbox/ondemand/providers/provider.py#L105

It doubly reads into the buffer, in one of the two cases only.

Reply:
I'm workin on it.